gateway/ln-gateway/src/bin/cln_extension.rs

```diff
@@ -279,29 +289,98 @@ pub enum ClnExtensionError {
     Error(#[from] anyhow::Error),
 }
 
+/// BOLT 4: https://github.com/lightning/bolts/blob/master/04-onion-routing.md#failure-messages
+/// 2002 error code reports general temporary failure of this processing node.
+fn temp_node_failure() -> serde_json::Value {
+    serde_json::json!({
+        "result": "fail",
+        "failure_message": "2002"
+    })
+}
+
 /// Functional structure to filter intercepted HTLCs into subscription streams.
 /// Used as a CLN plugin
 struct ClnHtlcInterceptor {
     subscriptions:
         Mutex<HashMap<u64, mpsc::Sender<Result<SubscribeInterceptHtlcsResponse, Status>>>>,
+    pub outcomes: Mutex<HashMap<sha256::Hash, oneshot::Sender<serde_json::Value>>>,
 }
 
 impl ClnHtlcInterceptor {
     fn new() -> Self {
         Self {
             subscriptions: Mutex::new(HashMap::new()),
+            outcomes: Mutex::new(HashMap::new()),
         }
     }
 
-    async fn intercept_htlc(&self, _payload: HtlcAccepted) -> serde_json::Value {
-        unimplemented!("TODO: Implement HTLC filtering using short channel ids")
+    async fn intercept_htlc(&self, payload: HtlcAccepted) -> serde_json::Value {
+        let short_channel_id = payload.onion.short_channel_id;
+        let htlc_expiry = payload.htlc.cltv_expiry;
+
+        if let Some(subscription) = self.subscriptions.lock().await.get(&short_channel_id) {
```

Comment:
Can we hold this lock for a bit shorter by restructuring the code? Right now it will be held up to 30s and block every other intercepted HTLC. @dpc is that assessment correct? I'm not sure why the lock would stick around, but if it didn't I don't see how that code would work without a call to `Sender::clone()`.

How I see it:

```rust
self.subscriptions.lock() // impl Future<Output = MutexGuard<'_, HashMap<ShortChannelId, Sender>>>
    .await // MutexGuard<'_, HashMap<ShortChannelId, Sender>>
    .get(&short_channel_id) // &Sender
```